Ignore the “title” argument to history.pushState()/history.replaceState()

sideshowbarker opened the pull request:
#### 4ef4b65d33f45734ad3c6cbc7f2fe0dda17051bc
<pre>
Ignore the “title” argument to history.pushState()/history.replaceState()
<a href="https://bugs.webkit.org/show_bug.cgi?id=223190">https://bugs.webkit.org/show_bug.cgi?id=223190</a>

Reviewed by Chris Dumez and Michael Catanzaro.

This change causes WebKit to ignore the second argument to history.pushState()
and history.replaceState(), making WebKit consistent with the current behavior
in Gecko and Blink, and consistent with the current spec.

That argument previously set the title for the history item, but in
<a href="https://github.com/whatwg/html/pull/6482">https://github.com/whatwg/html/pull/6482</a> was removed from the spec –
which now at <a href="https://html.spec.whatwg.org/#the-history-interface">https://html.spec.whatwg.org/#the-history-interface</a>:dom-history-pushstate
defines the argument as “unused”, and specifies no processing
requirements for it.

Otherwise, without this change, the WebKit behavior is inconsistent with
the current behavior in Gecko and Blink, and inconsistent with the
current requirements defined in the spec.

Note that this change also reverts <a href="https://commits.webkit.org/259367@main">https://commits.webkit.org/259367@main</a> —
because, since the title is being removed from history items, and no
longer used, there’s no longer any need to truncate it, nor any need for
the associated test.

* LayoutTests/loader/stateobjects/pushstate-size-expected.txt:
* LayoutTests/loader/stateobjects/pushstate-size-iframe-expected.txt:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::HistoryItem::reset):
(WebCore::HistoryItem::title const): Deleted.
(WebCore::HistoryItem::alternateTitle const): Deleted.
(WebCore::HistoryItem::setAlternateTitle): Deleted.
(WebCore::HistoryItem::setTitle): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didChangeTitle):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::initializeItem):
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
(WebCore::HistoryController::setCurrentItemTitle): Deleted.
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/History.h:
(WebCore::History::pushState):
(WebCore::History::replaceState):
* Source/WebCore/page/History.idl:
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const): Deleted.
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
(WKBackForwardListItemCopyTitle):
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem title]):
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.cpp:
(webkit_back_forward_list_item_get_title):
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.h.in:
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::toBackForwardListItemState):
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKitLegacy/mac/History/HistoryPropertyList.mm:
(HistoryPropertyListWriter::writeHistoryItem):
* Source/WebKitLegacy/mac/History/WebHistory.mm:
(-[WebHistoryPrivate visitedURL:withTitle:]):
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem initWithURLString:title:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithURLString:title:displayTitle:lastVisitedTimeInterval:]):
(-[WebHistoryItem _visited]):
(-[WebHistoryItem title]): Deleted.
(-[WebHistoryItem setAlternateTitle:]): Deleted.
(-[WebHistoryItem alternateTitle]): Deleted.
(-[WebHistoryItem setTitle:]): Deleted.
(-[WebHistoryItem _visitedWithTitle:]): Deleted.
* Source/WebKitLegacy/mac/History/WebHistoryItemInternal.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::setTitle):
* Tools/MiniBrowser/gtk/BrowserWindow.c:
(browserWindowCreateBackForwardMenu):
* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardListTests.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestBackForwardList.cpp:
(testBackForwardListNavigation):
(testWebKitWebViewSessionState):

Canonical link: <a href="https://commits.webkit.org/273650@main">https://commits.webkit.org/273650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05c7bff7c4383600a473a5fcfbffe80488f05d79

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38953 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32573 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37452 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17621 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12209 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31248 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12818 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32135 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11269 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40199 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/30540 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32882 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37203 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9355 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35296 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13188 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8222 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12311 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestBackForwardList.cpp:
```diff
@@ -69,11 +69,10 @@ class BackForwardListTest: public WebViewTest {
     {
         g_assert_nonnull(item);
         g_assert_cmpstr(webkit_back_forward_list_item_get_uri(item), ==, uri);
-        g_assert_cmpstr(webkit_back_forward_list_item_get_title(item), == , title);
```
Instead of removing this line, let's verify that it returns... whatever it returns, presumably nullptr?

sideshowbarker replied:
```
        g_assert_cmpstr(webkit_back_forward_list_item_get_title(item), == , title);
```
> Instead of removing this line, let's verify that it returns... whatever it returns, presumably nullptr?

Restored it and changed it to expect the empty string.

mcatanzaro reviewed (changes requested):
This looks good to me, assuming Apple is willing to do it. The impact is browsers can no longer show page titles in the back/forward list and have to show URLs instead. I checked Epiphany and found that WebKit was already just filling in URLs for the titles, though, so I'm not sure this will make much difference in practice.

Since we are changing the behavior of public APIs, we should also deprecate `webkit_back_forward_list_item_get_title()`, `WKBackForwardListItemCopyTitle`, and the title property of `WKBackForwardListItem`, and document what their new return values (NULL? empty string?)

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -868,8 +857,9 @@ void FrameLoader::HistoryController::updateCurrentItem()
-void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& stateObject, const String& title, const String& urlString)
+void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& stateObject, const String& unused, const String& urlString)
 {
+    UNUSED_PARAM(unused);
```
Please drop.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -868,8 +857,9 @@ void FrameLoader::HistoryController::updateCurrentItem()
 }
 
-void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& stateObject, const String& title, const String& urlString)
+void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& stateObject, const String& unused, const String& urlString)
```
Please omit parameter name instead of using `unused`. This is what we usually do in WebKit.

sideshowbarker replied:
> Please omit parameter name instead of using `unused`. This is what we usually do in WebKit.

OK, thanks — made that change throughout.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -910,16 +899,16 @@ void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& s
 }
 
-void FrameLoader::HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject, const String& title, const String& urlString)
+void FrameLoader::HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject, const String& unused, const String& urlString)
```
ditto.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -910,16 +899,16 @@ void FrameLoader::HistoryController::pushState(RefPtr<SerializedScriptValue>&& s
-void FrameLoader::HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject, const String& title, const String& urlString)
+void FrameLoader::HistoryController::replaceState(RefPtr<SerializedScriptValue>&& stateObject, const String& unused, const String& urlString)
 {
+    UNUSED_PARAM(unused);
```
ditto.

cdumez commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -86,8 +85,8 @@ class FrameLoader::HistoryController {
 
-    void pushState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
-    void replaceState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
+    void pushState(RefPtr<SerializedScriptValue>&&, const String& unused, const String& url);
```
ditto.

cdumez commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -86,8 +85,8 @@ class FrameLoader::HistoryController {
-    void pushState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
-    void replaceState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
+    void pushState(RefPtr<SerializedScriptValue>&&, const String& unused, const String& url);
+    void replaceState(RefPtr<SerializedScriptValue>&&, const String& unused, const String& url);
```
ditto.

cdumez commented on Source/WebCore/page/History.h:
```diff
@@ -65,14 +65,14 @@ class History final : public ScriptWrappable, public RefCounted<History>, public
 
-    ExceptionOr<void> pushState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString);
-    ExceptionOr<void> replaceState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString);
+    ExceptionOr<void> pushState(RefPtr<SerializedScriptValue>&& data, const String& unused, const String& urlString);
```
ditto.

cdumez commented on Source/WebCore/page/History.h:
```diff
@@ -65,14 +65,14 @@ class History final : public ScriptWrappable, public RefCounted<History>, public
-    ExceptionOr<void> pushState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString);
-    ExceptionOr<void> replaceState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString);
+    ExceptionOr<void> pushState(RefPtr<SerializedScriptValue>&& data, const String& unused, const String& urlString);
+    ExceptionOr<void> replaceState(RefPtr<SerializedScriptValue>&& data, const String& unused, const String& urlString);
```
ditto.

cdumez commented on Source/WebCore/page/History.h:
```diff
@@ -92,14 +92,16 @@ class History final : public ScriptWrappable, public RefCounted<History>, public
 };
 
-inline ExceptionOr<void> History::pushState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString)
+inline ExceptionOr<void> History::pushState(RefPtr<SerializedScriptValue>&& data, const String& unused, const String& urlString)
```
ditto.

cdumez commented on Source/WebCore/page/History.h:
```diff
@@ -92,14 +92,16 @@ class History final : public ScriptWrappable, public RefCounted<History>, public
 }
 
-inline ExceptionOr<void> History::replaceState(RefPtr<SerializedScriptValue>&& data, const String& title, const String& urlString)
+inline ExceptionOr<void> History::replaceState(RefPtr<SerializedScriptValue>&& data, const String& unused, const String& urlString)
```
ditto.

cdumez commented on Source/WebKitLegacy/mac/History/WebHistory.mm:
```diff
@@ -291,8 +291,6 @@ - (WebHistoryItem *)visitedURL:(NSURL *)url withTitle:(NSString *)title
         BOOL itemWasInDateCaches = [self removeItemFromDateCaches:entry.get()];
         ASSERT_UNUSED(itemWasInDateCaches, itemWasInDateCaches);
-
-        [entry _visitedWithTitle:title];
```
I think this looks risky. This method was setting the title sure, but it was also setting the last visited time. The last visited time is no longer getting set now, which seems like an unnecessary behavior change?

I suggest we name this to `_visited`

sideshowbarker replied:
```
        [entry _visitedWithTitle:title];
```

> I suggest we name this to `_visited`

OK, made that change

cdumez commented on Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
```diff
@@ -369,12 +349,6 @@ - (NSPoint)scrollPoint
-- (void)_visitedWithTitle:(NSString *)title
-{
-    core(_private)->setTitle(title);
-    _private->_lastVisitedTime = [NSDate timeIntervalSinceReferenceDate];
```
I suggest with keep this method, rename it to `_visited` and still update `_lastVisitedTime`.

sideshowbarker replied:
> I suggest with keep this method, rename it to `_visited` and still update `_lastVisitedTime`.

OK, made that change

cdumez reviewed (changes requested):
I'm OK with the behavior change. However, I'd like to make sure we keep updating the historyItem's lastVisitedTime in WebKitLegacy before I approve this.

mcatanzaro commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -86,8 +85,8 @@ class HistoryController {
-    void pushState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
-    void replaceState(RefPtr<SerializedScriptValue>&&, const String& title, const String& url);
+    void pushState(RefPtr<SerializedScriptValue>&&, const String&, const String& url);
+    void replaceState(RefPtr<SerializedScriptValue>&&, const String&, const String& url);
```
I think we can remove the parameters here instead of leaving them unused, right? They only have to remain in the History object itself?

sideshowbarker replied:
> I think we can remove the parameters here instead of leaving them unused, right? They only have to remain in the History object itself?

Given that they’re still required parameters in the web-exposed JS API — per the IDL and the method signatures at https://html.spec.whatwg.org/#the-history-interface:dom-history-pushstate — then, unless there’s some special way to handle it in the generated bindings code, don’t we still need to include the parameters in the C++ sources?

And just to be clear about the way they’re used and [documented](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState#javascript) for frontend code: developers normally just pass in the empty string as the 2nd arg:

```js
const state = { page_id: 1, user_id: 5 };
const url = "hello-world.html";
history.pushState(state, "", url);
```
That is, the JS calls still require that arg to be given (though giving it the empty string is just the normal convention for doing the minimal thing required; it could of course be any string — `"foo"` or whatever — since implementations aren’t required to actually do anything with it).

mcatanzaro replied:
> Given that they’re still required parameters in the web-exposed JS API — per the IDL and the method signatures at https://html.spec.whatwg.org/#the-history-interface:dom-history-pushstate — then, unless there’s some special way to handle it in the generated bindings code, don’t we still need to include the parameters in the C++ sources?

In the History object, absolutely yes. But that's not what I'm commenting on here.

There's probably no reason for HistoryController to keep these parameters, and for the History object to pass empty args along to HistoryController, right? Let's remove them wherever possible.

sideshowbarker replied:
> There's probably no reason for HistoryController to keep these parameters, and for the History object to pass empty args along to HistoryController, right? Let's remove them wherever possible.

I see now — yes, of course. (Thanks for your patience, and sorry for my confusion. I guess I had it stuck in my mind that the diff we were looking at there was for the History object  — despite it being clear from your review comment that this was about the HistoryController code…)

mcatanzaro commented on Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.h:
```diff
@@ -40,9 +40,9 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
+/*! @abstract Deprecated. Always returns the empty string.
  */
-@property (nullable, readonly, copy) NSString *title;
+@property (nullable, readonly, copy) NSString *title DEPRECATED_ATTRIBUTE;
```
I'm not too familiar with the Cocoa APIs, but I think you need to use `WK_API_DEPRECATED` here rather than `DEPRECATED_ATTRIBUTE`. But that requires specifying the macOS and iOS versions that the API is deprecated in. You might need to ask an Apple developer for help with this.

I'm not even sure where `DEPRECATED_ATTRIBUTE` comes from. It doesn't seem to be defined anywhere, and is used only by libwebrtc. Not sure how you found that.

mcatanzaro replied:
@cdumez can you provide an example of how to deprecate this API, please?

mcatanzaro commented on Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
```diff
@@ -41,6 +41,7 @@ WKURLRef WKBackForwardListItemCopyURL(WKBackForwardListItemRef itemRef)
+// This function is deprecated and now always returns the empty string.
 WKStringRef WKBackForwardListItemCopyTitle(WKBackForwardListItemRef itemRef)
 {
     return WebKit::toCopiedAPI(toImpl(itemRef)->title());
```
There's no need to leave comments saying the API is deprecated. Those have no effect; nobody will ever notice them. Instead, use `WK_C_API_DEPRECATED` on the declaration of this function in the header file, which is what will trigger `-Wdeprecated-declaration`.

Please also change it to just directly return the empty string. The goal is to comprehensively remove the title property wherever possible (everywhere except where it is exposed to the web or exposed in API).

sideshowbarker replied:
```cpp
WKStringRef WKBackForwardListItemCopyTitle(WKBackForwardListItemRef itemRef)
{
    return WebKit::toCopiedAPI(toImpl(itemRef)->title());
``` 
> Please also change it to just directly return the empty string.

I changed it to `return 0` — since the return type of the function is `WKStringRef`. Or do you mean I should also change the return type of the function? (so that I can make it return the empty string)

mcatanzaro replied:
No, certainly don't change the return type of the function.

I think this function should return a non-NULL/nonzero WKStringRef that contains an empty string.

mcatanzaro replied:
Actually I see Chris is saying this is fine if changed to `nullptr`. He has more experience with these C APIs than I do, so let's trust his answer instead.

cdumez replied:
I actually like Michael's idea better. Let's return an empty string here. It is less likely to cause backward compatibility issues. Thanks for catching it @mcatanzaro !

mcatanzaro commented on Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
```diff
@@ -48,6 +48,7 @@ - (NSURL *)URL
+// This property is deprecated and now always returns the empty string.
 - (NSString *)title
 {
     if (!_item->title())
```
* No need for the deprecation comment
 * Please directly return an empty string instead of returning `_item->title()`

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.cpp:
```diff
@@ -121,6 +121,8 @@ const gchar* webkit_back_forward_list_item_get_uri(WebKitBackForwardListItem* li
+ *
  * Returns: the page title of @list_item or %NULL
  *    when the title is empty.
  */
```
```suggestion
 * Returns: an empty string
 *
 * Deprecated: 2.44
 */
```

The implementation should be just g_return_val_if_fail() followed by returning an empty string.

You'll also need to change the `WEBKIT_API` on the declaration to `WEBKIT_DEPRECATED` in WebKitBackForwardListItem.h.in.

mcatanzaro reviewed (changes requested):
I'd like to remove as much of the title property as possible. Please also remove `WebBackForwardListItem::title` to make sure that code still using it cannot compile.

cdumez commented on Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
```diff
@@ -43,7 +43,7 @@ WKURLRef WKBackForwardListItemCopyURL(WKBackForwardListItemRef itemRef)
 WKStringRef WKBackForwardListItemCopyTitle(WKBackForwardListItemRef itemRef)
 {
-    return WebKit::toCopiedAPI(toImpl(itemRef)->title());
+    return 0;
```
`return nullptr;`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.h:
```diff
@@ -40,7 +40,7 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
-/*! @abstract The title of the webpage represented by this item.
+/*! @abstract Deprecated. Always returns nil.
  */
 @property (nullable, readonly, copy) NSString *title;
```
Please use:
```
@property (nullable, readonly, copy) NSString *title WK_API_DEPRECATED("No longer supported", macOS(10.10, WK_MAC_TBA), iOS(8.0, WK_IOS_TBA))
```

cc @achristensen07

cdumez reviewed (approved):
LGTM

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.cpp:
```diff
@@ -121,20 +121,14 @@ const gchar* webkit_back_forward_list_item_get_uri(WebKitBackForwardListItem* li
  *
  * Obtain the title of the item.
```
```suggestion
 * Since 2.44, page titles are no longer stored in history. This function now returns an empty string.
```

mcatanzaro commented on Tools/MiniBrowser/gtk/BrowserWindow.c:
```diff
@@ -156,9 +156,7 @@ static void webViewURIChanged(WebKitWebView *webView, GParamSpec *pspec, Browser
-    const char *title = webkit_web_view_get_title(webView);
-    if (!title)
-        title = defaultWindowTitle;
+    const char *title = defaultWindowTitle;
```
Oops, sorry I didn't notice earlier. I was about to submit this for merge queue, too.

This shouldn't be removed. Here we set the window title from the web view's title, which has nothing to do with the back/forward list. Remove this change, please.

sideshowbarker replied:
```
    const char *title = webkit_web_view_get_title(webView);
    if (!title)
        title = defaultWindowTitle;
```

> This shouldn't be removed. Here we set the window title from the web view's title, which has nothing to do with the back/forward list. Remove this change, please.

Thanks for catching that — fixed now

sideshowbarker committed:
Ignore the “title” argument to history.pushState()/history.replaceState()
https://bugs.webkit.org/show_bug.cgi?id=223190

Reviewed by Chris Dumez and Michael Catanzaro.

This change causes WebKit to ignore the second argument to history.pushState()
and history.replaceState(), making WebKit consistent with the current behavior
in Gecko and Blink, and consistent with the current spec.

That argument previously set the title for the history item, but in
https://github.com/whatwg/html/pull/6482 was removed from the spec –
which now at https://html.spec.whatwg.org/#the-history-interface:dom-history-pushstate
defines the argument as “unused”, and specifies no processing
requirements for it.

Otherwise, without this change, the WebKit behavior is inconsistent with
the current behavior in Gecko and Blink, and inconsistent with the
current requirements defined in the spec.

Note that this change also reverts https://commits.webkit.org/259367@main —
because, since the title is being removed from history items, and no
longer used, there’s no longer any need to truncate it, nor any need for
the associated test.

* LayoutTests/loader/stateobjects/pushstate-size-expected.txt:
* LayoutTests/loader/stateobjects/pushstate-size-iframe-expected.txt:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::HistoryItem):
(WebCore::HistoryItem::reset):
(WebCore::HistoryItem::title const): Deleted.
(WebCore::HistoryItem::alternateTitle const): Deleted.
(WebCore::HistoryItem::setAlternateTitle): Deleted.
(WebCore::HistoryItem::setTitle): Deleted.
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::create):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didChangeTitle):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::initializeItem):
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
(WebCore::HistoryController::setCurrentItemTitle): Deleted.
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/History.h:
(WebCore::History::pushState):
(WebCore::History::replaceState):
* Source/WebCore/page/History.idl:
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::url const):
(WebKit::WebBackForwardListItem::title const): Deleted.
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.cpp:
(WKBackForwardListItemCopyTitle):
* Source/WebKit/UIProcess/API/C/WKBackForwardListItemRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.h:
* Source/WebKit/UIProcess/API/Cocoa/WKBackForwardListItem.mm:
(-[WKBackForwardListItem title]):
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.cpp:
(webkit_back_forward_list_item_get_title):
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.h.in:
* Source/WebKit/WebProcess/WebCoreSupport/SessionStateConversion.cpp:
(WebKit::toBackForwardListItemState):
(WebKit::applyFrameState):
(WebKit::toHistoryItem):
* Source/WebKitLegacy/mac/History/HistoryPropertyList.mm:
(HistoryPropertyListWriter::writeHistoryItem):
* Source/WebKitLegacy/mac/History/WebHistory.mm:
(-[WebHistoryPrivate visitedURL:withTitle:]):
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem initWithURLString:title:lastVisitedTimeInterval:]):
(-[WebHistoryItem initWithURLString:title:displayTitle:lastVisitedTimeInterval:]):
(-[WebHistoryItem _visited]):
(-[WebHistoryItem title]): Deleted.
(-[WebHistoryItem setAlternateTitle:]): Deleted.
(-[WebHistoryItem alternateTitle]): Deleted.
(-[WebHistoryItem setTitle:]): Deleted.
(-[WebHistoryItem _visitedWithTitle:]): Deleted.
* Source/WebKitLegacy/mac/History/WebHistoryItemInternal.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::setTitle):
* Tools/MiniBrowser/gtk/BrowserWindow.c:
(browserWindowCreateBackForwardMenu):
* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardListTests.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestBackForwardList.cpp:
(testBackForwardListNavigation):
(testWebKitWebViewSessionState):

Canonical link: https://commits.webkit.org/273650@main
